File: Java/5. Longest Palindromic Substring.java
class Solution {
    
    private Integer maxLen = 0;
    private Integer maxLeft = 0, maxRight = 0;
    
    private void isPalindrome(Integer left, Integer right, String s) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            if (right - left + 1 > maxLen) {
                maxLen = right - left + 1;
                maxLeft = left;
                maxRight = right;
            }
            left -= 1;
            right += 1;
        }        
    }
    
    public String longestPalindrome(String s) {        
        if (s.length() == 0) return "";
        for (int i = 0; i < s.length(); i++) {
            isPalindrome(i, i, s);
            if (i < s.length() - 1 && s.charAt(i) == s.charAt(i + 1))
                isPalindrome(i, i + 1, s);            
        }
        return s.substring(maxLeft, maxRight + 1);
    }
}

File: Python/5. Longest Palindromic Substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        def expand(s, left, right):
            l = left
            r = right
            while l >= 0 and r < len(s) and s[l] == s[r]:
                l -= 1
                r += 1
            return r - l -1;
        start = end = 0
        for i in range(len(s)):
            len1 = expand(s, i, i)
            len2 = expand(s, i, i + 1)
            length = max(len1, len2)            
            if (length > end - start):
                start = i - (length - 1) // 2
                end = i + length//2
        return s[start:end + 1]
